### PR TITLE
evalengine: Improve handling of datetime / time with precision and add more functions

### PR DESCRIPTION
**Pull request by dbussink**
This PR contains a bunch of improvements for the datetime / time handling we've been working on in the evalengine. It builds on top of #12892 which should be merged first and then this can be rebased. It's best reviewed also on a commit by commit basis, but these are the main changes.

#### Ensure we can better match exact changes

Right now we have non-exact matching for datetime / time values since we want to test functions like `NOW()`. The problem is that always enabling this hides real bugs such as a differences in precision. This is due to the fact that we need a high tolerate of 1 second to avoid flakes in CI since GitHub Actions can stall for a significant amount of time. 

With the change, we instead make non-exact matching optional and only enable it for `NOW()` and exact match for all other datetime / time related functions such as casting and formatting of those values. This exposes a number of bugs in our handling of these objects already. This is subsequently fixed.

#### Improve the parser for explicit precision

When parsing datetime / time values, we often have an explicitly requested precision (such as in a `CAST(... AS datetime(6))`. This needs to be handled and the value needs to be rounded properly as well. Functions like `CONVERT_TZ` on the other hand, keep whatever the input precision is, so that's something we also keep. This means parsing functions both have a requested precision (or `-1` when keeping whatever the input has). And they return the found precision.

#### Fix how we check for compiler errors

Before we would do the stack size check when compiling a function as a return for translating an expression. This was hiding bugs, because a failure in translation in our test framework is treated as "unsupported function" and isn't an error state for the tests.

Instead, we need to run the stack check as part of the compilation and set `compilerErr` properly for it, so that the tests can catch bugs here. These bugs were there and they are also fixed.

#### Implement a slew of additional function

With `DATE_FORMAT` we had a lot of formatter options, but a lot of those are also available as separate functions. This adds support for those functions as well. Some caveats:

- We don't honor [default_week_format](https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_default_week_format) for the `WEEK` function at this point and assume the default of mode `0` when not given.
- Week modes `4` through `7` are not yet supported, just `0` through `3` which are the ones that `DATE_FORMAT` supports as well.

## Related Issue(s)

Follow up to #12892 & #12815 and part of #9647

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on the CI
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vtgate/evalengine/compiler_asm.go [hunk 1]
```diff
@@ -3077,13 +3117,302 @@ func (asm *assembler) Fn_RandomBytes() {
...
+			env.vm.stack[env.vm.sp-1] = env.vm.arena.newEvalInt64(0)
+		case 1, 2, 3:
+			env.vm.stack[env.vm.sp-1] = env.vm.arena.newEvalInt64(1)
+		case 4, 5, 6:
+			env.vm.stack[env.vm.sp-1] = env.vm.arena.newEvalInt64(2)
+		case 7, 8, 9:
+			env.vm.stack[env.vm.sp-1] = env.vm.arena.newEvalInt64(3)
+		case 10, 11, 12:
+			env.vm.stack[env.vm.sp-1] = env.vm.arena.newEvalInt64(4)
+		}
```
Comment (vmg) [inline go/vt/vtgate/evalengine/compiler_asm.go:3285]: This is definitely worth lifting into a method in `DateTime`.

File: go/vt/vtgate/evalengine/compiler_asm.go [hunk 2]
```diff
@@ -3077,13 +3117,302 @@ func (asm *assembler) Fn_RandomBytes() {
...
+				week = 0
+			}
+			env.vm.stack[env.vm.sp-2] = env.vm.arena.newEvalInt64(int64(week))
+		case 2:
+			_, week := arg.dt.Date.SundayWeek()
+			env.vm.stack[env.vm.sp-2] = env.vm.arena.newEvalInt64(int64(week))
+		case 3:
+			_, week := arg.dt.Date.ISOWeek()
+			env.vm.stack[env.vm.sp-2] = env.vm.arena.newEvalInt64(int64(week))
+		}
```
Comment (vmg) [inline go/vt/vtgate/evalengine/compiler_asm.go:3345]: Same, method in `Date`

File: go/vt/vtgate/evalengine/compiler_asm.go [hunk 3]
```diff
@@ -3077,13 +3117,302 @@ func (asm *assembler) Fn_RandomBytes() {
...
+		arg := env.vm.stack[env.vm.sp-2].(*evalTemporal)
+		mode := env.vm.stack[env.vm.sp-1].(*evalInt64)
+
+		switch mode.i {
+		case 0, 2:
+			year, week := arg.dt.Date.SundayWeek()
+			env.vm.stack[env.vm.sp-2] = env.vm.arena.newEvalInt64(int64(year*100 + week))
+		case 1, 3:
+			year, week := arg.dt.Date.ISOWeek()
+			env.vm.stack[env.vm.sp-2] = env.vm.arena.newEvalInt64(int64(year*100 + week))
```
Comment (vmg) [inline go/vt/vtgate/evalengine/compiler_asm.go:3413]: Also method in `Date`

**Review by vmg — APPROVED** [review]

File: go/mysql/datetime/format_test.go [hunk 4]
```diff
@@ -45,7 +45,7 @@ func TestFormattingFromMySQL(t *testing.T) {
 
 	for _, tc := range cases {
 		t.Run(tc.timestamp, func(t *testing.T) {
-			dt, ok := ParseDateTime(tc.timestamp)
+			dt, _, ok := ParseDateTime(tc.timestamp, -1)
```
Comment (mattlord) [inline go/mysql/datetime/format_test.go:48]: IMO it's worth adding a consts in the datetime package for -1,3,6... so that it's more obvious what the mean. Looks like you already did do that at least for `DefaultPrecision`, we're just not using them everywhere. The other ones are more obvious, but IMO still worth adding:
```
dt, _, ok := ParseDateTime(tc.timestamp, DefaultPrecision)

dt, _, ok := ParseDateTime(tc.timestamp, SecondPrecision)

dt, _, ok := ParseDateTime(tc.timestamp, MillisecondPrecision)

dt, _, ok := ParseDateTime(tc.timestamp, MicrosecondPrecision)

dt, _, ok := ParseDateTime(tc.timestamp, NanosecondPrecision)
```

Don't feel strongly about that, so up to you.

File: go/mysql/datetime/timeparts.go [hunk 5]
```diff
@@ -70,7 +71,16 @@ func (tp *timeparts) toDateTime() (DateTime, bool) {
 			second:     uint8(tp.sec),
 			nanosecond: uint32(tp.nsec),
 		},
-	}, true
+	}
+
+	l := prec
+	if prec < 0 {
```
Comment (mattlord) [inline go/mysql/datetime/timeparts.go:77]: Is it worth explicitly checking for -1 in case e.g. -2 might mean something else in the future? Not sure... just came to mind.

File: go/mysql/datetime/types.go [hunk 6]
```diff
@@ -212,6 +245,69 @@ func (d Date) SundayWeek() (int, int) {
...
+		return week
+	case 2:
+		_, week := d.SundayWeek()
+		return week
+	case 3:
+		_, week := d.ISOWeek()
+		return week
+	case 4, 5, 6, 7:
+		// TODO
+		return 0
```
Comment (mattlord) [inline go/mysql/datetime/types.go:272]: This (and YearWeek) feels like we're explicitly allowing for unexpected behavior. I do see that you already have a separate PR open for this though.

**Review by mattlord — APPROVED** [review]
I like it. Thank you! ❤️ I think this will address a number of failures that I've seen in the CI over time.

File: go/mysql/datetime/types.go (hunk 6 above)
Comment (dbussink) [reply go/mysql/datetime/types.go:272]: @mattlord Yeah, looking into this in a follow up to fix these and add support for them.

File: go/mysql/datetime/timeparts.go (hunk 5 above)
Comment (dbussink) [reply go/mysql/datetime/timeparts.go:77]: Right, but in that case things don't work anywhere so not sure how that helps. It trades one undefined behavior for another and I don't really see a strong preference for either anything way.

File: go/mysql/datetime/types.go (hunk 6 above)
Comment (dbussink) [reply go/mysql/datetime/types.go:272]: @mattlord See https://github.com/vitessio/vitess/pull/12916
